docs/hugo/content/architecture-decision-records/ADR-2022-01-Reconciler-Extensions.md

```diff
@@ -0,0 +1,123 @@
+
+It's quite likely that some extension points will need to import our generated resource types to allow per-version manipulation. This rules out referencing extensions from the generated resource types.
+
+We don't want extension points to reimplement functionality already present in the generic reconciler as this runs the risk of inconsistencies (especially if we change the generic reconciler itself), and introduces a significant overhead to implementation. Instead, the extension points will provide an opportunity to modify the information already held. An extension point that does nothing should have the same effect as one that hasn't been implemented.
+
+## Decision
+
+To provide a stable implementation location for extensions, we'll introduce a dedicated package alongside the existing versioned packages for each supported service. This package will provide extension support for all supported resource versions. For example, alongside the existing **batch** packages `v1alpha1api20210101`, and `v1alpha1api20210101storage`, we will generate `extensions`.
+
+Into the `extensions` package we will codegen a skeleton framework. For each resource there will be a separate extension type, suffixed with `Extensions`. For example, for the **compute** resources `Disk`, `VirtualMachine`, and `VirtualMachineScaleSet` we will create the extension types `DiskExtensions`, `VirtualMachineExtensions`, and `VirtualMachineScaleSetExtensions`.
+
+We'll enhance the existing registration file (`controller_resources_gen.go`) to also provide registration of extensions. They will be indexed by the GVK of each resource, allowing easy lookup by the generic reconciler.
+
+For each supported extension, we'll define an interface. Implementers will assert type compatibility with this extension interface to ensure they have the correct method signature. The generic reconciler will identify available extensions by testing for the presence of the interface.
+
+### Example
+
+To allow customization of error handling for Azure Redis, we'll introduce an extension point, which would look something like this:
+
+We create the required extension interface in `genruntime`:
+
+``` go
+type ErrorClassifierExtension interface {
+    // Classify the provided error, returning details about the classification.
+    // cloudError is the error returned from ARM.
+    // apiVersion is the ARM API version used for the request
+    // next is the function to call for standard classification behaviour
+    ClassifyError(
+        cloudError *genericarmclient.CloudError, 
+        apiVersion string, 
+        next func(*genericarmclient.CloudError) (*CloudErrorDetails, error)) (*CloudErrorDetails, error)
+}
+```
+
+In the `cache/extensions` package, a host type for the extension point will be generated in `redis_extensions_gen.go`:
+
+```
+type RedisExtensions struct{}
+```
+
+**TODO**: Should this have any members?
```

Comment:
Probably not? They're really just containers for extension methods right? Probably clearer to pass whatever data we need in the method signatures than through here, I would think.

Reply:
Unless we have some sort of `MakeExtensionHandler` interface that produces these and _thats_ what we're registering?

Reply:
Agreed.